source.go

```diff
@@ -15,81 +38,216 @@ type Source struct {
-	// last record that was successfully processed, Source should therefore
-	// start producing records after this position. The context passed to Open
-	// will be cancelled once the plugin receives a stop signal from Conduit.
+func (s *Source) Open(ctx context.Context, pos opencdc.Position) error {
+	sdk.Logger(ctx).Info().Msg("Opening DynamoDB Source...")
+
+	cfg, err := config.LoadDefaultConfig(ctx,
+		config.WithRegion(s.config.AWSRegion),
+		config.WithCredentialsProvider(aws.NewCredentialsCache(credentials.NewStaticCredentialsProvider(s.config.AWSAccessKeyID, s.config.AWSSecretAccessKey, ""))),
+	)
+	if err != nil {
+		return fmt.Errorf("error creating AWS session: %w", err)
+	}
+
+	s.dynamoDBClient = dynamodb.NewFromConfig(cfg)
+	s.streamsClient = dynamodbstreams.NewFromConfig(cfg)
+	s.lastPositionRead = pos
+
+	partitionKey, sortKey, err := s.getKeyNamesFromTable(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting key names from table: %w", err)
+	}
+	err = s.prepareStream(ctx)
+	if err != nil {
+		return err
+	}
+
+	p, err := position.ParseRecordPosition(pos)
+	if err != nil {
+		return fmt.Errorf("error parssing position: %w", err)
+	}
+
+	// create the needed iterator
+	var itr Iterator
+	if s.config.SkipSnapshot {
+		itr, err = iterator.NewCDCIterator(ctx, s.config.Table, partitionKey, sortKey, s.config.PollingPeriod, s.streamsClient, s.streamArn, p)
+		if err != nil {
+			return fmt.Errorf("error creating CDC iterator: %w", err)
+		}
+	} else {
+		itr, err = iterator.NewCombinedIterator(ctx, s.config.Table, partitionKey, sortKey, s.config.PollingPeriod, s.dynamoDBClient, s.streamsClient, s.streamArn, p)
+		if err != nil {
+			return fmt.Errorf("error creating combined iterator: %w", err)
+		}
+	}
+	s.iterator = itr
 	return nil
 }
 
-func (s *Source) Read(_ context.Context) (opencdc.Record, error) {
-	// Read returns a new Record and is supposed to block until there is either
-	// a new record or the context gets cancelled. It can also return the error
-	// ErrBackoffRetry to signal to the SDK it should call Read again with a
-	// backoff retry.
-	// If Read receives a cancelled context or the context is cancelled while
-	// Read is running it must stop retrieving new records from the source
-	// system and start returning records that have already been buffered. If
-	// there are no buffered records left Read must return the context error to
-	// signal a graceful stop. If Read returns ErrBackoffRetry while the context
-	// is cancelled it will also signal that there are no records left and Read
-	// won't be called again.
-	// After Read returns an error the function won't be called again (except if
-	// the error is ErrBackoffRetry, as mentioned above).
-	// Read can be called concurrently with Ack.
-	return opencdc.Record{}, nil
+func (s *Source) Read(ctx context.Context) (opencdc.Record, error) {
+	sdk.Logger(ctx).Info().Msg("Reading records from DynamoDB...")
```

Comment:
`Info` could be a little bit too verbose, WDYT? Maybe `Trace`?

Reply:
yep agreed